### PR TITLE
Decouple emailservice from app package 

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
This PR moves email service into a separate package for the sake of app modularization.

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: api4/team_local.go [hunk 1]
```diff
@@ -126,7 +128,14 @@ func localInviteUsersToTeam(c *Context, w http.ResponseWriter, r *http.Request)
 		if len(goodEmails) > 0 {
 			err := c.App.Srv().EmailService.SendInviteEmails(team, "Administrator", "mmctl "+model.NewId(), goodEmails, *c.App.Config().ServiceSettings.SiteURL)
 			if err != nil {
-				c.Err = err
+				switch {
+				case errors.Is(err, email.NoRateLimiterError):
```
Comment (agnivade) [inline api4/team_local.go:132]: Any reason we are not printing the error here too?

File: app/email/email.go [hunk 2]
```diff
@@ -480,43 +413,41 @@ func (es *EmailService) sendMfaChangeEmail(email string, activated bool, locale,
...
+		return SetupRateLimiterError
 	}
 
 	if rateLimited {
-		return model.NewAppError("SendInviteEmails",
-			"app.email.rate_limit_exceeded.app_error", map[string]interface{}{"RetryAfter": result.RetryAfter.String(), "ResetAfter": result.ResetAfter.String()},
-			fmt.Sprintf("user_id=%s, team_id=%s, retry_after_secs=%f, reset_after_secs=%f",
-				senderUserId, team.Id, result.RetryAfter.Seconds(), result.ResetAfter.Seconds()),
-			http.StatusRequestEntityTooLarge)
+		mlog.Error("rate limit exceeded", mlog.Duration("RetryAfter", result.RetryAfter), mlog.Duration("ResetAfter", result.ResetAfter), mlog.String("user_id", senderUserId),
```
Comment (agnivade) [inline app/email/email.go:438]: Any reason we are logging and returning the error? Can we not log it at the call-site?

File: app/email/errors.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+package email
+
+import "github.com/pkg/errors"
+
+var (
+	CreateEmailTokenError = errors.New("could not create token")
+	NoRateLimiterError    = errors.New("the rate limit could not be found")
+	SetupRateLimiterError = errors.New("the rate limiter could not be set")
+	RateLimitExceedError  = errors.New("the rate limit is exceeded")
```
Comment (agnivade) [inline app/email/errors.go:12]: Nit: `RateLimitExceededError`

File: services/users/profile_picture.go [hunk 4]
```diff
@@ -10,20 +10,85 @@ import (
...
 	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/shared/filestore"
 	"github.com/mattermost/mattermost-server/v5/utils/fileutils"
 )
 
 const (
 	imageProfilePixelDimension = 128
 )
 
+func (us *UserService) GetProfileImage(user *model.User) ([]byte, bool, error) {
```
Comment (agnivade) [inline services/users/profile_picture.go:29]: unrelated?

File: services/users/profile_picture.go [hunk 5]
```diff
@@ -10,20 +10,85 @@ import (
...
+	if *us.config().FileSettings.DriverName == "" {
+		img, err := us.GetDefaultProfileImage(user)
+		if err != nil {
+			return nil, false, err
+		}
+		return img, false, nil
+	}
+
+	path := "users/" + user.Id + "/profile.png"
```
Comment (agnivade) [inline services/users/profile_picture.go:38]: Let's use `path.Join`

**Review by agnivade — COMMENTED** [review]
Very thorough work! Purple sectors all over!

Just left some minor comments.

**Review by AGMETEOR — APPROVED** [review]

File: app/email/email.go (hunk 2 above)
Comment (isacikgoz) [reply app/email/email.go:439]: This function is called several times and TBH I was not comfortable doing so but I think it may be simpler to keep this way as it would need to be changed for all occurrences if it ever changes.

File: api4/team_local.go (hunk 1 above)
Comment (isacikgoz) [reply api4/team_local.go:132]: Answered below.

File: services/users/profile_picture.go (hunk 4 above)
Comment (isacikgoz) [reply services/users/profile_picture.go:29]: It was being used in email service :)

File: app/email/email.go (hunk 2 above)
Comment (agnivade) [reply app/email/email.go:439]: ok

**Review by agnivade — APPROVED** [review]
Neat work!

**Comment by isacikgoz** [comment]
@saturninoabril some testing around email stuff should work e.g. automated e2e ones :)

**Review by saturninoabril — APPROVED** [review]
Tested and passed - https://automation-dashboard.vercel.app/cycles/69bdda8d-dd48-42b7-b288-f852fedc8056
